Asset dto

stark-dev opened the pull request:
New fty::Asset class
- conversion to/from JSON
- conversion to/from fty-proto
- duplicated zhash_to_map/map_to_zhash to remove MlmUtils dependency
- added explicit cxxtools dep in project.xml
- lot of diff in project regen due to new version of generation scripts

etnKlendathu commented on include/fty_asset_dto.h:
```diff
@@ -0,0 +1,184 @@
+    =========================================================================
+*/
+
+#ifndef FTY_ASSET_DTO_H_INCLUDED
+#define FTY_ASSET_DTO_H_INCLUDED
```
as header guard you could use #pragma once

etnKlendathu commented on src/fty_asset_dto.cc:
```diff
@@ -0,0 +1,717 @@
+#include <zhash.h>
+
+// for fty-proto conversion
+static fty::HashMap zhashToMap(zhash_t *hash)
+{
+    fty::HashMap map;
+    char *item = (char *)zhash_first(hash);
+    while(item) {
+        const char * key = zhash_cursor(hash);
+        const char * val = (const char *)zhash_lookup(hash,key);
+        if( key && val ) map[key] = val;
+        item = (char *)zhash_next(hash);
+    }
+    return map;
+}
```
Try out:
for(auto* item = zhash_first(map); item; item=zhash_next(map)) {
    map.emplace(zhash_cursor(map), static_cast<const char*>(item));
}

stark-dev replied:
Sure, the only issue may be if either/both key or/and value are NULL, but shouldn't happen

etnKlendathu commented on src/fty_asset_dto.cc:
```diff
@@ -0,0 +1,717 @@
+
+    Asset::Asset()
+    {
+        m_id = "";
```
="" is not needed, std::string aways initialized as ""

stark-dev replied:
👍

etnKlendathu commented on src/fty_asset_dto.cc:
```diff
@@ -0,0 +1,717 @@
+    {
+        m_id = "";
+        m_assetStatus = Status_Unknown;
+        m_assetType = persist::TUNKNOWN;
```
Such initialization could be in static class initilaization (i.e. in header)
`private:
    MyEnum m_enum = MyEnum::DefValue;
`

etnKlendathu commented on src/fty_asset_dto.cc:
```diff
@@ -0,0 +1,717 @@
+        if (si.findMember("ext"))
+        {
+            const cxxtools::SerializationInfo extSi = si.getMember("ext");
+            for (const auto element : extSi)
```
copy of element is needed here?

stark-dev replied:
not sure, you suggest to use smart pointer?

etnKlendathu replied:
`const auto&` element :)

etnKlendathu commented on include/fty_asset_dto.h:
```diff
@@ -0,0 +1,184 @@
+
+namespace fty
+{
+    using HashMap = std::map<std::string, std::string>;
```
Och, this is dengerous, you defined a internal type in fty namespace in header.

stark-dev replied:
How should I deal with this? I mean, placing the definition outside the namespace in the header means that anyone including this header will import that definition. Shall I just drop it and use standard map?

etnKlendathu replied:
Just define it in class.

etnKlendathu commented on src/fty_asset_dto.cc:
```diff
@@ -0,0 +1,717 @@
+        std::string tmpString;
+        HashMap tmpMap;
+        // status
+        if (!si.findMember("status"))
+        {
+            throw std::runtime_error("Member status does not exist");
+        }
+        si.getMember("status") >>= tmpInt;
```
Maybe something like this? And... maybe keys-strings should be defined as constants...
```
auto getByKey = [&](const std::string& key) {
    if (!si.findMember(key.c_str())) {
       throw std::runtime_error("Member " + key + " does not exist");
    }
    return si.getMemeber(key.c_str());
};

getByKey("status") >>= tmpInt;
getByKey("next") >>= tmpStr;
```

stark-dev replied:
I like it!

etnKlendathu commented on src/fty_asset_dto.cc:
```diff
@@ -0,0 +1,717 @@
+
+        zhash_t *ext = mapToZhash(asset.getExt());
+
+        fty_proto_set_aux(proto, &aux);
```
mmm... for example from fty_proto header about `fty_proto_set_aux`
/// Set the aux field, tranferring ownership from caller

Is zhash_destroy(&aux) works? No fail?

stark-dev replied:
it's working, yes.. no clue on how this should work and who should handle the creation.
I would prefer to do the new externally and pass the pointer already initialized, this way you do create & destroy at the same level

etnKlendathu replied:
Mmmm... main thing was: if you transfer ownership to different context then very dangerous to destroy this pointer because it should be destroyed in different context.

etnKlendathu replied:
Try comment destroy here and check memleak in valgrind?
I just not sure...

stark-dev replied:
I have memory leak when removing that part :/

stark-dev commented on include/fty_asset_dto.h:
```diff
@@ -0,0 +1,184 @@
+    using HashMap = std::map<std::string, std::string>;
+
+    /// List of valid asset statuses
+    enum AssetStatus : uint8_t
```
Could be removed at some point, but may be changed to enum class as well :)

etnKlendathu commented on include/fty_asset_dto.h:
```diff
@@ -140,21 +139,26 @@ namespace fty
         uint8_t     m_priority;
         /// ext map storage (asset-specific values)
-        HashMap     m_ext;
+        std::map<std::string, std::string>     m_ext;
```
nono... sorry for bad explanation :(
````
class Asset {
public:
     using HashMap = std::map....;
.....
     HashMap map();
}
````

etnKlendathu commented on include/fty_asset_dto.h:
```diff
@@ -140,21 +139,26 @@ namespace fty
         public:
-        Asset();
-        ~Asset();
+        Asset() :
+            m_assetStatus(AssetStatus::Status_Unknown),
+            m_assetType(TYPE_UNKNOWN),
+            m_assetSubtype(SUB_UNKNOWN),
+            m_priority(5)
+        {};
```
I wanted a static initialialization... just remove ctor.... and in private section:
```
AssetStatus m_assetStatus = AssetStatus::Status_Unknown;
std::string m_assetType = TYPE_UNKNOWN;
.....
```

etnKlendathu commented on include/fty_asset_dto.h:
```diff
@@ -140,21 +139,26 @@ namespace fty
-        const std::string getParentId() const { return m_parentId; };
-        const uint8_t     getPriority() const { return m_priority; };
-        const HashMap     getExt() const { return m_ext; };
+        const std::string                           getId() const { return m_id; };
```
btw, here you could return `const std::string&`

etnKlendathu commented on include/fty_asset_dto.h:
```diff
@@ -33,17 +32,16 @@
         Status_Unknown = 0,
         Status_Active,
         Status_Nonactive
     };
```
if `enum class` then `Status_` prefix in enum values is overkill, isnt't it? :)

boricj commented on include/fty_asset_dto.h:
```diff
@@ -0,0 +1,187 @@
+#include <fty_common.h>
+#include <ftyproto.h>
+
+#include "cxxtools/serializationinfo.h"
+#include "cxxtools/jsondeserializer.h"
```
These should be system includes (`#include <...>`).

boricj commented on include/fty_asset_dto.h:
```diff
@@ -0,0 +1,187 @@
+        /// direct parent iname
+        std::string m_parentId;
+        /// priority 1..5 (1 is most, 5 is least)
+        uint8_t     m_priority;
```
Any reason why you're not using `int` here? You're not trying to represent a byte or a value between 0 and 255 and struct padding will not yield space savings in memory anyway.

stark-dev replied:
Not really, just replicating what was done in ExtendedAsset class, it's stored as uint8_t there. Explanation is that priority is 1 to 5. If it's unnecessary, then I'll go for int

boricj commented on include/fty_asset_dto.h:
```diff
@@ -0,0 +1,187 @@
+    {
+        private:
+        /// internal identification string (iname)
+        std::string m_id;
```
Canonical ID for an asset in database is an integer. The `type-number` storage format used is partially redundant with `m_assetSubtype`. Store the ID and subtype separately, build the `iname` on demand whenever requested.

boricj commented on include/fty_asset_dto.h:
```diff
@@ -0,0 +1,187 @@
+        /// friendly name (hist. ename) provided by user
+        std::string m_friendlyName;
+        /// direct parent iname
+        std::string m_parentId;
```
Like the `m_id` comment, store parent ID and parent asset subtype separately.

boricj commented on src/fty_asset_dto.cc:
```diff
@@ -0,0 +1,686 @@
+
+    void Asset::setAssetSubtype(const std::string & assetSubtype)
+    {
+        m_assetSubtype = assetSubtype;
```
Consequence of using `iname` as identifier: the `m_id` property is not updated when you update the asset subtype.

boricj commented on src/fty_asset_dto.cc:
```diff
@@ -0,0 +1,686 @@
+
+    bool Asset::operator== (const Asset &asset) const
+    {
+        if(m_id != asset.m_id)
+        {
+            return false;
+        }
+        if(m_assetStatus != asset.m_assetStatus)
+        {
+            return false;
+        }
+        if(m_assetType != asset.m_assetType)
+        {
+            return false;
+        }
+        if(m_assetSubtype != asset.m_assetSubtype)
+        {
+            return false;
+        }
+        if(m_friendlyName != asset.m_friendlyName)
+        {
+            return false;
+        }
+        if(m_parentId != asset.m_parentId)
+        {
+            return false;
+        }
+        if(m_parentId != asset.m_parentId)
+        {
+            return false;
+        }
+        if(m_priority != asset.m_priority)
+        {
+            return false;
+        }
```
Replace with a single boolean expression with "or" operator.

boricj commented on src/fty_asset_dto.cc:
```diff
@@ -0,0 +1,686 @@
+        std::string tmpString;
+        std::map<std::string, std::string> tmpMap;
+
+        auto getByKey = [&](const std::string& key)
+        {
+            if(!si.findMember(key.c_str())) {
+                throw std::runtime_error("Member " + key + " does not exist");
+            }
+            return si.getMember(key.c_str());
+        };
```
Won't `gerMember()` already throw an exception if the key does not exist?

stark-dev replied:
I will check, thanks

etnKlendathu commented on src/fty_asset_dto.cc:
```diff
@@ -0,0 +1,704 @@
+
+void fty_asset_dto_test(bool verbose)
+{
+    printf (" * fty_asset_dto: ");
```
Are you really want to use printf here? :) In code above you used cout :)

etnKlendathu commented on include/fty_asset_dto.h:
```diff
@@ -0,0 +1,183 @@
+    };
+
+    // WARNING keep consistent with DB table t_bios_asset_element_type
+    static constexpr const char *TYPE_UNKNOWN                         = "unknown";         // 0
```
Somehow leaked from me, where this heap of constexpr stuff is used?

etnKlendathu commented on src/fty_asset_dto.cc:
```diff
@@ -0,0 +1,704 @@
+
+namespace fty
+{
+    std::map<AssetStatus, std::string> AssetStatusToProto =
+    {
+        {AssetStatus::Unknown, "unknown"},
+        {AssetStatus::Active, "active"},
+        {AssetStatus::Nonactive, "nonactive"}
+    };
+
+    std::map<std::string, fty::AssetStatus> ProtoToAssetStatus = 
+    {
+        {"unknown", AssetStatus::Unknown},
+        {"active", AssetStatus::Active},
+        {"nonactive", AssetStatus::Nonactive}
+    };
```
If I should implement enum -> string -> enum then I will implement as function with value check (and do not use `at` operator below, which fires just wrong index exception).
You could write it as:
```
std::string toString(AssetStatus status) {
     static std::map AssetStatusToProto = {
     };
     if (value = find(); value != map.end()) {
         throw std:::runtime_error(Wrong value);
     }
     return value;
}
Or better to use a switch here... which will help you to handle new enum values case.
```

etnKlendathu commented on src/fty_asset_dto.cc:
```diff
@@ -0,0 +1,704 @@
+
+        cxxtools::SerializationInfo &extSi = si.addMember("ext");
+        extSi.setCategory (cxxtools::SerializationInfo::Object);
+        for (const auto keyValue : asset.getExt())
```
const auto& keyValue

etnKlendathu commented on src/fty_asset_dto.cc:
```diff
@@ -0,0 +1,704 @@
+        si.getMember("location") >>= tmpString;
+        asset.setParentId(tmpString);
+
+        // id (if missing assign -1)
+        if (!si.findMember("id"))
+        {
+            tmpInt = -1;
+        }
+        else
+        {
+            si.getMember("id") >>= tmpInt;
+        }
+        asset.setId(tmpInt);
```
Exists bool function getMember
```
bool found = si.getMember("id", tmpInt));
asset.setId(found ? tmpInt : -1);
```

stark-dev replied:
nice ;)

etnKlendathu commented on src/fty_asset_dto.cc:
```diff
@@ -0,0 +1,704 @@
+        asset.setId(tmpInt);
+
+        // ext map
+        tmpMap.clear();
```
This is not a c lang :)
Declare variable in place where you use it, if you do it, you will see what clear here is uneccesary.

etnKlendathu commented on src/fty_asset_dto.cc:
```diff
@@ -0,0 +1,704 @@
+                        // only DB is interested in read_only attribute
+                        if (innerKey != "read_only")
+                        {
+                            std::string value;
+                            innerElement >>= value;
+                            tmpMap[innerKey] = value;
```
Why are you need a `value` var here? :)

etnKlendathu commented on src/fty_asset_dto.cc:
```diff
@@ -0,0 +1,704 @@
+        zhash_t *aux = zhash_new();
+        zhash_autofree (aux);
+
+        zhash_insert(aux, "priority", (void*) std::to_string(asset.getPriority()).c_str());
```
Very picky note, in c++ (some_type*) conversion is marked as old style and did not recomented to use. Just use static_cast, reinterpret_cast etc here...

etnKlendathu commented on src/fty_asset_dto.cc:
```diff
@@ -0,0 +1,704 @@
+{
+    zhash_t *hash = zhash_new ();
+    for (const auto & i :map) {
+        zhash_insert (hash, i.first.c_str (), (void*) i.second.c_str());
```
Very picky note, in c++ (some_type*) conversion is marked as old style and did not recomented to use. Just use static_cast, reinterpret_cast etc here..

stark-dev commented on include/fty_asset_dto.h:
```diff
@@ -0,0 +1,183 @@
+        const std::string         internalName() const;
+
+        // getters
+        const int                 getId() const;
```
Right, I forgot to remove, it was string before ;)

stark-dev commented on src/fty_asset_dto.cc:
```diff
@@ -0,0 +1,704 @@
+            m_priority     != asset.m_priority
+        )
+        {
+            std::cout << m_id           << " : " <<  asset.m_id << std::endl;
+            std::cout << (int)m_assetStatus  << " : " << (int)asset.m_assetStatus << std::endl;
+            std::cout << m_assetType    << " : " <<  asset.m_assetType << std::endl;
+            std::cout << m_assetSubtype << " : " <<  asset.m_assetSubtype << std::endl;
+            std::cout << m_friendlyName << " : " <<  asset.m_friendlyName << std::endl;
+            std::cout << m_parentId     << " : " <<  asset.m_parentId << std::endl;
+            std::cout << m_parentId     << " : " <<  asset.m_parentId << std::endl;
+            std::cout << m_priority     << " : " <<  asset.m_priority << std::endl;
```
Should't push on friday evening :/

stark-dev commented:
@boricj, @etnKlendathu: updated again the PR, addressed some of the issues you pointed out. Thanks for your comments.

@boricj: after discussion with @perrettecl, I reverted the id representation to a string. I know that this may lead to some inconsistency between asset subtype and internal name, but this asset class should be populated only from the db or from data on message bus, direct manipulation should be avoided (at least, as far as I understood). If you still see an issue, we may set up a quick discussion for this design. Thanks
